src/include/cache.i

```diff
@@ -199,49 +230,20 @@ __wt_cache_full_check(WT_SESSION_IMPL *session)
 	 * Threads operating on trees that cannot be evicted are ignored,
 	 * mostly because they're not contributing to the problem.
 	 */
-	if ((btree = S2BT_SAFE(session)) != NULL &&
-	    F_ISSET(btree, WT_BTREE_NO_EVICTION))
-		return (0);
-
-	/*
-	 * Only wake the eviction server the first time through here (if the
-	 * cache is too full).
-	 *
-	 * If the cache is less than 95% full, no work to be done.  If we are
-	 * at the API boundary and the cache is more than 95% full, try to
-	 * evict at least one page before we start an operation.  This helps
-	 * with some eviction-dominated workloads.
-	 */
-	WT_RET(__wt_eviction_check(session, &full, 1));
-	if (full < 95)
+	btree = S2BT_SAFE(session);
+	if (btree != NULL && F_ISSET(btree, WT_BTREE_NO_EVICTION))
 		return (0);
 
-	return (__wt_cache_wait(session, full));
-}
-
-/*
- * __wt_session_can_wait --
- *	Return if a session available for a potentially slow operation.
- */
-static inline int
-__wt_session_can_wait(WT_SESSION_IMPL *session)
-{
-	/*
-	 * Return if a session available for a potentially slow operation;
-	 * for example, used by the block manager in the case of flushing
-	 * the system cache.
-	 */
-	if (!F_ISSET(session, WT_SESSION_CAN_WAIT))
+	/* Check if eviction is needed. */
+	if (!__wt_eviction_needed(session, &pct_full))
 		return (0);
 
 	/*
-	 * LSM sets the no-cache-check flag when holding the LSM tree lock,
-	 * in that case, or when holding the schema lock, we don't want to
-	 * highjack the thread for eviction.
+	 * Some callers (those waiting for slow operations), will sleep if there
+	 * was no cache work to do. After this point, let them skip the sleep.
 	 */
-	if (F_ISSET(session,
-	    WT_SESSION_NO_CACHE_CHECK | WT_SESSION_LOCKED_SCHEMA))
-		return (0);
+	if (didworkp != NULL)
+		*didworkp = 1;
 
-	return (1);
+	return (__wt_cache_eviction_worker(session, busy, pct_full));
```

Comment:
"busy" threads only evict a single page before returning from that function, other threads evict 10 pages and then return.